events: Emit lease expiration events

swenson opened the pull request:
This is to gather feedback.

We might want to support other lease events, do different fields, or do these in a different way. And definitely add some tests.

This works by patching the lease expiration function so that it emit events when the lease timer fires.

I also investigated using a B-Tree to keep track of upcoming expirations, as it would probably be more scalable and allow us to fire events *before* leases expire, but it adds a fair amount of complexity and will require tuning.

These events look like this:

```json
{
  "id": "expire-auth/approle/login/hc16d231ec9efbe88e9c49fc5d691f2617879146a13bb9b6a90ddc37d6ef02a84-1694728738",
  "source": "https://vaultproject.io/",
  "specversion": "1.0",
  "type": "*",
  "data": {
    "event": {
      "id": "expire-auth/approle/login/hc16d231ec9efbe88e9c49fc5d691f2617879146a13bb9b6a90ddc37d6ef02a84-1694728738",
      "metadata": {
        "client_token_type": "default",
        "expire_time": "2023-09-14T14:58:58-07:00",
        "id": "auth/approle/login/hc16d231ec9efbe88e9c49fc5d691f2617879146a13bb9b6a90ddc37d6ef02a84",
        "last_renewal_time": "0001-01-01T00:00:00Z",
        "login_role": "example-dot-com",
        "namespace": "",
        "path": "auth/approle/login",
        "version": "0"
      },
      "entity_ids": [
        "auth/approle/login/hc16d231ec9efbe88e9c49fc5d691f2617879146a13bb9b6a90ddc37d6ef02a84"
      ]
    },
    "event_type": "core/lease-expiration"
  },
  "datacontentype": "application/cloudevents",
  "time": "2023-09-14T14:58:58.456021-07:00"
}
```

tomcf-hcp commented on vault/expiration.go:
```diff
@@ -325,6 +328,14 @@ func getNumExpirationWorkers(c *Core, l log.Logger) int {
+// alsoSendEvent returns an ExpireLeaseStrategy that sends an expiration event after executing the given ExpireLeaseStrategy.
+func alsoSendEvent(e ExpireLeaseStrategy) ExpireLeaseStrategy {
+	return func(ctx context.Context, m *ExpirationManager, leaseID string, namespace *namespace.Namespace) {
+		e(ctx, m, leaseID, namespace)
```
Are there any failure conditions for `e` that we should include/emit in the expiration event? Or don't send the event if e fails?
If not, are there any trade-offs between starting the go routine before or after `e`?

swenson replied:
I think the right thing to do is to *not* send the event if `e` panics, but there's not currently a way to know if it failed other than that.

The benefit to starting after is that if `e` does panic, then the event won't be sent.

The biggest disadvantage of this approach is that we have to fetch the lease from storage again when we send the event, which costs a little bit of performance. We could fix this by moving the event sending into the `e` function itself, but this would require modifying each of the `ExpireLeaseStrategy` implementations, which is more work to maintain. Or we could rework the `ExpireLeaseStrategy` interface so we could avoid the additional fetch of the lease.

ncabatoff replied:
We may want to draw a distinction between expiration and revocation.  In other words, what we're doing now in this PR is firing an event saying a lease expired, but ignoring whether or not the lease expiration succeeded, and whether or not the associated dynamic secrets, if any, got revoked.

ncabatoff commented:
One problem I see here is that leaseIDs are sensitive information: if you know the leaseID, you can revoke or renew it.  It's not unusual for the /sys/leases/revoke endpoint to be granted in policies, and sys/leases/renew is in the default policy.  Renewal could be used as part of an attack, e.g.: I don't have access to this crashed application's token, but I do have a valid token from elsewhere, and I want to make use of the dynamic db credential I stole from the app.

swenson commented:
> One problem I see here is that leaseIDs are sensitive information: if you know the leaseID, you can revoke or renew it. It's not unusual for the /sys/leases/revoke endpoint to be granted in policies, and sys/leases/renew is in the default policy. Renewal could be used as part of an attack, e.g.: I don't have access to this crashed application's token, but I do have a valid token from elsewhere, and I want to make use of the dynamic db credential I stole from the app.

Perfect. This is exactly the sort of feedback I was looking for. :)

swenson commented:
@ncabatoff I updated and standardized the lease event structure, and now emit them on creation, revocation, and renewal.

I believe I removed the sensitive data, but still have enough info for them to be actionable. Though, the `data_path` is not present, since I can't think of anything useful for it that doesn't involve the lease ID.

I also added a simple test.

Events look like:

```json
{
  "id": "c07bf537-14cd-ad14-4d64-fcb5de6cf295",
  "source": "https://vaultproject.io/",
  "specversion": "1.0",
  "type": "*",
  "data": {
    "event": {
      "id": "c07bf537-14cd-ad14-4d64-fcb5de6cf295",
      "metadata": {
        "client_token_type": "default",
        "expire_time": "2023-09-20T11:08:55-07:00",
        "last_renewal_time": "0001-01-01T00:00:00Z",
        "login_role": "example-dot-com",
        "namespace": "",
        "path": "auth/approle/login",
        "token_accessor": "LkFUFhSYEKaV8YTEtimPn3cL",
        "token_entity_id": "41e53bca-bb1b-4ca9-09a5-184de1f39ce2",
        "version": "1"
      },
      "entity_ids": [
        "hvs.G615l6Y7vFExM3XlxRn71scQ",
        "41e53bca-bb1b-4ca9-09a5-184de1f39ce2"
      ]
    },
    "event_type": "core/lease-create"
  },
  "datacontentype": "application/cloudevents",
  "time": "2023-09-20T11:08:54.185301-07:00"
}
```

```json
{
  "id": "7ebbbf41-f239-5b83-7f6f-a4bff055b2f8",
  "source": "https://vaultproject.io/",
  "specversion": "1.0",
  "type": "*",
  "data": {
    "event": {
      "id": "7ebbbf41-f239-5b83-7f6f-a4bff055b2f8",
      "metadata": {
        "client_token_type": "default",
        "expire_time": "2023-09-20T11:08:55-07:00",
        "last_renewal_time": "0001-01-01T00:00:00Z",
        "login_role": "example-dot-com",
        "namespace": "",
        "path": "auth/approle/login",
        "token_accessor": "LkFUFhSYEKaV8YTEtimPn3cL",
        "token_entity_id": "41e53bca-bb1b-4ca9-09a5-184de1f39ce2",
        "version": "1"
      },
      "entity_ids": [
        "hvs.G615l6Y7vFExM3XlxRn71scQ",
        "41e53bca-bb1b-4ca9-09a5-184de1f39ce2"
      ]
    },
    "event_type": "core/lease-revoke"
  },
  "datacontentype": "application/cloudevents",
  "time": "2023-09-20T11:08:55.187169-07:00"
}
```

benashz commented:
> One problem I see here is that leaseIDs are sensitive information: if you know the leaseID, you can revoke or renew it. It's not unusual for the /sys/leases/revoke endpoint to be granted in policies, and sys/leases/renew is in the default policy. Renewal could be used as part of an attack, e.g.: I don't have access to this crashed application's token, but I do have a valid token from elsewhere, and I want to make use of the dynamic db credential I stole from the app.

@ncabatoff  That's really good to know. So are you saying we should never include the leaseID in logs, events, etc., since those sources may leak out of the producer?

ncabatoff commented on vault/expiration.go:
```diff
@@ -1806,6 +1823,55 @@ func (m *ExpirationManager) deleteLockForLease(id string) {
 	m.lockPerLease.Delete(id)
 }
 
+func (m *ExpirationManager) sendEventByID(eventType logical.EventType, leaseID string) {
```
This doesn't seem to be used?

ncabatoff commented on vault/expiration.go:
```diff
@@ -1806,6 +1823,55 @@ func (m *ExpirationManager) deleteLockForLease(id string) {
+	}
+	data.Id = id
+	data.EntityIds = []string{le.ClientToken}
+	structMap := map[string]interface{}{
```
Should last_renewal_time only be populated if nonzero?  Should we include IssueTime (when it was created)?  What about RevokeErr?

ncabatoff commented on vault/expiration.go:
```diff
@@ -1806,6 +1823,55 @@ func (m *ExpirationManager) deleteLockForLease(id string) {
+		"client_token_type": le.ClientTokenType.String(),
+		"namespace":         ns.Path,
+	}
+	if le.Auth != nil {
```
My recollection is that in a leaseEntry, we commonly only preserve a handful of Auth fields: namespace, policies, and auth method.  We need the Auth structure to route renewals IIRC, and for some metrics we emit, but we try to minimize how much of it we keep around, for memory reasons.

swenson replied:
I understanding of this code is far from complete, but it looks like the only that `le.Auth` is trimmed is for in-memory storage in `m.irrevocable`, `m.nonexpiring`, and `m.pending`, but it looks like it is always put into storage with all the fields?

swenson replied:
(And so far, we are only using the `le.Auth` that comes from storage in these code paths.)

ncabatoff replied:
Hmm, you're right it looks like, I'm probably thinking of a big PR I never merged.  There are exceptions, like ControlGroupStore.createToken and wrapInCubbyHole, but in the main what you say is true as far as I can tell.

ncabatoff commented on vault/expiration_test.go:
```diff
@@ -424,6 +424,73 @@ func TestExpiration_TotalLeaseCount_WithRoles(t *testing.T) {
+	for {
+		select {
+		case <-timeout.C:
+			if expectedEvents != eventsCount.Load() {
```
Would this be better off checking for event types, to prevent the test breaking if we ever start emitting events for other reasons, and to ensure that we're not just getting say two createlease events instead of one create and one expire?

swenson replied:
Makes sense. I was trying to keep it simple, but you're right, we should check specific types.

swenson commented:
@ncabatoff and others, any other feedback on this, or other work I should do to get this ready for merging?

tomhjp commented on vault/expiration.go:
```diff
@@ -1806,6 +1822,67 @@ func (m *ExpirationManager) deleteLockForLease(id string) {
+	m.sendEvent(eventType, le)
+}
+
+func (m *ExpirationManager) sendEvent(eventType logical.EventType, le *leaseEntry) {
```
(This comment is all about understanding any further implications for a pre-existing issue, so is about discussion rather than suggesting any changes)

The `ExpirationManager` runs on every node that's not a DR secondary:

https://github.com/hashicorp/vault/blob/87376a1f5c676c9a54e1050b3e71e32c4ac92595/vault/core.go#L2392-L2398

However, clients of the events system are forwarded to the leader, so we aren't currently able to consume events from Perf secondaries, because they don't forward their events within the cluster. IIUC before this PR we would only be able to generate events on Perf secondaries if they have a local mount and thus accept writes to storage directly for that mount [citation needed from @ncabatoff]. Do lease events change the picture here at all?

tomhjp commented on vault/expiration.go:
```diff
@@ -1806,6 +1822,67 @@ func (m *ExpirationManager) deleteLockForLease(id string) {
+		ns = le.namespace
+	}
+	data.Id = id
+	data.EntityIds = []string{le.ClientToken}
```
Is `ClientToken` a secret (it very much sounds like it)? If so, we shouldn't put it in the event

swenson replied:
I thought it wasn't, for some reason, but upon further inspection, it appears to possibly be a secret. I'll remove it.

tomhjp reviewed (commented):
I'm not an expert on the ExpirationManager side, but broadly LGTM from the events side. Just 2 questions
